storage/range.go

```diff
@@ -420,11 +422,17 @@ func (r *Range) Desc() *proto.RangeDescriptor {
 	return (*proto.RangeDescriptor)(atomic.LoadPointer(&r.desc))
 }
 
-// SetDesc atomically sets the range's descriptor. This method should
-// be called in the context of having metaLock held, as is the case
-// for merging, splitting and updating the replica set.
-func (r *Range) SetDesc(desc *proto.RangeDescriptor) {
+// SetDesc atomically sets the range's descriptor. This method calls
+// ProcessRangeDescriptorUpdate() to make the range manager handle the
+// descriptor update. Note that ProcessRangeDescriptorUpdate()
+// acquires the metaLock.
```

Comment:
ProcessRangeDescriptorUpdate acquires the store lock, not the metaLock. All the operations that call SetDesc should hold the metaLock, although this mechanism doesn't really look correct at the moment since the metaLock is only acquired on the leader, and if leadership is lost the new leader could start a new metaLock operation while the previous transaction is still ongoing. Add a `TODO(bdarnell)` to revisit the metaLock.